### ISSUE TITLE
Minute corruption of unrepresentable float

### ISSUE THREAD

**Issue by g-andrade** [rurban/Cpanel-JSON-XS#154]
I encountered a corner case and I'm not sure whether it fits the spec on numerical precision and accuracy - **decoding the `0.3` numerical value**, which has no precise representation in binary floating point.

Its _closest_ representable value is `2.99999999999999988898e-01`, but JSON-XS decodes it as `0.30000000000000004` [instead](https://github.com/maxmind/MaxMind-DB/issues/72). I suspect the root cause to be somewhere in the `json_atof` function or one of its dependencies.

Here's a small script that reproduces the issue:
```perl
use Cpanel::JSON::XS qw( decode_json );

my $literal_float = 0.3;
my $json_object = decode_json('{"value": 0.3}');
my $json_float = $json_object->{value};

my $difference = $literal_float - $json_float;
print "$difference\n"
```
It outputs `-5.55111512312578e-17`. Is this expected?

**Comment by rurban** [comment]
A round-trip loss of < 1E16 is expected, but a bit unfortunate. Have to check how normal stringification deals with that problem.

**Comment by rurban** [comment]
Nope, my initial guess was wrong. JSON::PP::number() does the following trick to round the stringification: 0.3 - ("0.3"/1.0) => 0. i.e. division by 1.0

**Comment by rurban** [comment]
This can be easily fixed since 5.30 only, which has the first reliable Perl_strtod, being independent on LC_NUMERIC. Before Perl_strtod was available since 5.21.4.

With HAVE_BAD_POWL being set, the accuracy is also correct, using strtold on both conversions. (freebsd 10 only)

**Comment by rurban** [comment]
Thanks for this very important finding! Fixed in 4.16

**Comment by g-andrade** [comment]
Cool, thanks for the quick response :)
